### ISSUE TITLE
Add support for use of environment variable names in YAML configuration file paths

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#274]
From Make a better plan for relative paths in example-datasets [example-datasets#27](https://github.com/riboviz/example-datasets/issues/27):

Defining in configuration YAML
```
example_relative_path: ${EXAMPLE}/example.txt
```
Updating `prep_riboviz.nf`:
```
print("=====\n")
example="${EXAMPLE}"
print(example)
print("\n")
print(params.example_relative_path)
print("\n")
new_path = params.example_relative_path.replace('${EXAMPLE}', example)
print(new_path)
print("\n")
print("=====\n")
```
Running,
```console
$ EXAMPLE=123 nextflow run prep_riboviz.nf -params-file vignette/vignette_config.yaml -ansi-log false --publish_index_tmp --validate_only
```
gives:
```
=====
123
${EXAMPLE}/example.txt
123/example.txt
=====
i.e. the token `'${EXAMPLE}'` is replaced by the value of that environment variable.
```

Allow environment variables specifying paths to files to be optionally specified in values of YAML configuration file properties. If present these should be replaced with the value of the corresponding environment variable. 

Token replacement need cares. We don't want a situation where a user has a parameter `some_key: ${RIBOVIZ_PATH}/somefile.txt` and they provide no value for `RIBOVIZ_PATH` and the value of `some_key` ends up as `/somefile.txt`, when we'd want `somefile.txt`. If no value is provided for `RIBOVIZ_PATH` then this could default to ".", to give `./somefile.txt`. Experimentation is required to ensure the behaviour is as expected.

Possible tests

| `config.yaml` directory | paths in `config.yaml` parameters | Nextflow run location | Notes |
| - | - | - | - |
| `riboviz/` | relative to current directory (`RIBOVIZ_INPUT` or `RIBOVIZ_DATA` unused) | `riboviz/` | For example, run `vignette/vignette_config.yaml` as at present |
| `example/`, sibling folder of `riboviz/` or of its parents (i.e. not a descendant) | relative to current directory as above | `example/` |  For example, copy `vignette/` and `inputs/` into a sibling of `riboviz/`, run Nextflow within the sibling as described in [Invoking the workflow from outwith the RiboViz home directory](https://github.com/riboviz/riboviz/blob/develop/docs/user/prep-riboviz-run-nextflow.md#invoking-the-workflow-from-outwith-the-riboviz-home-directory) |
|  `example/` as above | use `RIBOVIZ_INPUT` or `RIBOVIZ_DATA` | `example/` | `RIBOVIZ_INPUT` and `RIBOVIZ_DATA` provided as environment variables |

This has implications for using the regression test framework and any other component (e.g. `riboviz.tools.count_reads`) that uses path values from configuration YAML. These too would need to be updated to do the environment variable substitution if the associated tokens are present in configuration values.

**Comment by mikej888** [comment]
@ewallace, @FlicAnderson Do you want two or three environment variables? What are your preferred names `RIBOVIZ_<N>`?

**Comment by FlicAnderson** [comment]
I'm not sure I've fully understood how to avoid the "/somefile.txt" vs "somefile.txt" issue, so apologies if my thoughts as follows would proliferate exactly that sort of issue...

I agree as discussed in the last dev meeting that some of the yeast files in `riboviz/data/` folder could be duplicated to `example-datasets` repo as they are files required to run the datasets, and aren't huge, so could be included.  We'd need to put the equivalent of those files into example-datasets anyway for other organisms, so it makes sense to include the S. cerevisiae ones too. 
 
Doing this might then remove a need for a third environment variable in 'full-size real data' examples, but I imagine we'd need to have 3 anyway to cover the vignette case? In which case, I suppose we could just advise users that in many cases, RIBOVIZ_DATA and RIBOVIZ_ORGANISM_FILES would be the same, but RIBOVIZ_INPUT (potentially RIBOVIZ_SAMPLE_FILES?) would potentially be different? 

But, I suppose this doesn't account for the RIBOVIZ_PATH location variable you mentioned as well, which might bring us up to 4 potential environment variables?

So in a 'run dataset G-Sc_2018' example from Eddie, this would go: 

RIBOVIZ_PATH = `$HOME/riboviz/riboviz`
(this is where my repo sits)

RIBOVIZ_SAMPLE_FILES = `/exports/eddie/scratch/$USER/G-Sc_2018/`
(this covers .fastq file `input/SRR8375039.fastq.gz` which is big and I don't want on my home directory and keep in scratch)

RIBOVIZ_ORGANISM_FILES = `$HOME/riboviz/example-datasets/fungi/saccharomyces/`
( e.g. this covers files from example-datasets repo: 
 - `orf_fasta_file`: `annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.fa`
 - `orf_gff_file`: `annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.gff3`
 - `rrna_fasta_file`: `contaminants/Saccharomyces_cerevisiae_yeast_rRNA_R64-1-1.fa`
- `sample_sheet`: `annotation/barcodes/Gupta_2018_tRNA_Modification_Carbon_Nitrogen_Metabolism_RPF_9-samples_CDS_w_250utrs_config_barcode.tsv`
- `asite_disp_length_file`: `other/yeast_standard_asite_disp_length.txt`  (NOTE: this is a file currently held in `data/`, if we're duplicating to example-datasets repo, we should replace 'other' with some useful descriptive :smile: )
- `features_file`: `other/yeast_features.tsv` (another current `data/` file to duplicate to ex-dt repo)
)

RIBOVIZ_DATA: not used in this example, as we've copied those files to the example-datasets repo, covered by RIBOVIZ_ORGANISM_FILES

YAML: 
```
asite_disp_length_file: ${RIBOVIZ_ORGANISM_FILES}/other/yeast_standard_asite_disp_length.txt 
dir_in: ${RIBOVIZ_SAMPLE_FILES}/input # Input directory
dir_out:  ${RIBOVIZ_SAMPLE_FILES}/output # Output directory
features_file: ${RIBOVIZ_ORGANISM_FILES}/other/yeast_features.tsv
orf_fasta_file: ${RIBOVIZ_ORGANISM_FILES}/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.fa 
orf_gff_file: ${RIBOVIZ_ORGANISM_FILES}/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.gff3
rrna_fasta_file: ${RIBOVIZ_ORGANISM_FILES}/contaminants/Saccharomyces_cerevisiae_yeast_rRNA_R64-1-1.fa
sample_sheet: ${RIBOVIZ_ORGANISM_FILES}/annotation/barcodes/Gupta_2018_tRNA_Modification_Carbon_Nitrogen_Metabolism_RPF_9-samples_CDS_w_250utrs_config_barcode.tsv 
```
Does this make sense according to what you're proposing  @mikej888?

I think it might work quite well, and is flexible if we just need to decide which of these environment variables are best suited for each yaml parameter.  

For another example, if I was setting up a new simdata_umi vignette I might have:
```
RIBOVIZ_PATH = $HOME/riboviz/riboviz
RIBOVIZ_SAMPLE_FILES = $HOME/riboviz/ribovz/vignette/
RIBOVIZ_ORGANISM_FILES = $HOME/riboviz/ribovz/vignette/input
RIBOVIZ_DATA = $HOME/riboviz/ribovz/data/
```
I guess maybe depending on your level of splitting/lumping, these things could be nested, like:
RIBOVIZ_PATH = $HOME/riboviz/riboviz
RIBOVIZ_SAMPLE_FILES = ${RIBOVIZ_PATH}/vignette/

With a yaml like this:
```
dir_in: ${RIBOVIZ_DATA}/simdata/
dir_out: ${RIBOVIZ_SAMPLE_FILES}/simdata_umi_output
orf_fasta_file: ${RIBOVIZ_ORGANISM_FILES}/yeast_YAL_CDS_w_250utrs.fa
```
I hope this thinking aloud is helpful/relevant?  Feel free to point out ways where I've made things overly complicated or misunderstood - paths are my nemesis it seems!

**Comment by mikej888** [comment]
> I'm not sure I've fully understood how to avoid the "/somefile.txt" vs "somefile.txt" issue, so apologies if my thoughts as follows would proliferate exactly that sort of issue...

`somefile.txt` is relative to the current directory whereas '/somefile.txt' would try and find `somefile.txt` in the root directory `/`.

**Comment by mikej888** [comment]
`RIBOVIZ_PATH` was used by way of example, to be overridden by your preferred choices :-) which are `RIBOVIZ_SAMPLE_FILES`, `RIBOVIZ_ORGANISM_FILES` and `RIBOVIZ_DATA`. Your examples are helpful and relevant and show how this would work in practice in terms of what files would be accessed via what environment variable.

**Comment by FlicAnderson** [comment]
> `RIBOVIZ_PATH` was used by way of example, to be overridden by your preferred choices :-) 

Ooops, yes. I guess nextflow knows where it's running riboviz from :grin:

**Comment by mikej888** [comment]
Updated example of accessing and replacing variables using environment variables:
```
example = System.getenv('EXAMPLE')
print("EXAMPLE: " + example + "\n")
if (! example)
{
    example = "."
}
print("EXAMPLE (post-null check): " + example + "\n")
print("Original params.example_relative_path: " + params.example_relative_path + "\n")
// Use single-quoted string in replacement to ensure EXAMPLE is	not
// evaluated. We want as-is replacement.
new_path = params.example_relative_path.replace('${EXAMPLE}', example)
print("Updated params.example.relative_path: " + new_path + "\n")
```
Examples, using parameter provided at command-line (and using single quotes to prevent bash evaluating environment variable). An alternative would be to add:
```
example_relative_path: ${EXAMPLE}/example.txt
```
to the YAML.

```console
$ EXAMPLE=/home/ubuntu/riboviz nextflow run prep_riboviz.nf -params-file vignette/simdata_umi_config.yaml --validate_only --example_relative_path '${EXAMPLE}/someDataFile.txt'
EXAMPLE: /home/ubuntu/riboviz
EXAMPLE (post-null check): /home/ubuntu/riboviz
Original params.example_relative_path: ${EXAMPLE}/someDataFile.txt
Updated params.example.relative_path: /home/ubuntu/riboviz/someDataFile.txt

$ nextflow run prep_riboviz.nf -params-file vignette/simdata_umi_config.yaml --validate_only --example_relative_path '${EXAMPLE}/someDataFile.txt'
EXAMPLE: null
EXAMPLE (post-null check): .
Original params.example_relative_path: ${EXAMPLE}/someDataFile.txt
Updated params.example.relative_path: ./someDataFile.txt
```
Examples with parameters with no `${EXAMPLE}` substring:
```console
$ EXAMPLE=/home/ubuntu/riboviz nextflow run prep_riboviz.nf -params-file vignette/simdata_umi_config.yaml --validate_only --example_relative_path 'someDataFile.txt'
EXAMPLE: /home/ubuntu/riboviz
EXAMPLE (post-null check): /home/ubuntu/riboviz
Original params.example_relative_path: someDataFile.txt
Updated params.example.relative_path: someDataFile.txt

$ nextflow run prep_riboviz.nf -params-file vignette/simdata_umi_config.yaml --validate_only --example_relative_path 'someDataFile.txt'
EXAMPLE: null
EXAMPLE (post-null check): .
Original params.example_relative_path: someDataFile.txt
Updated params.example.relative_path: someDataFile.txt
```

**Comment by mikej888** [comment]
Commit d01b165 on [yaml-environment-vars-274](https://github.com/riboviz/riboviz/tree/yaml-environment-vars-274)

* Replaced `file(params.<PARAMETER>)` calls with `file(replace_tokens(params.<PARAMETER>, riboviz_env_paths)` for `<PARAMETER>`s:
```
asite_disp_length_file
codon_positions_file
features_file
orf_fasta_file
orf_gff_file
rrna_fasta_file
t_rna_file
```
* Extracted `params.<PARAMETER>` into `<PARAMETER>` local variables with value `replace_tokens(params.<PARAMETER>, riboviz_env_paths)` for directory `<PARAMETER>`s:
```
dir_in
dir_index
dir_out
dir_tmp
```

**Comment by mikej888** [comment]
@FlicAnderson In your example you include `sample_sheet`:

> sample_sheet: annotation/barcodes/Gupta_2018_tRNA_Modification_Carbon_Nitrogen_Metabolism_RPF_9-samples_CDS_w_250utrs_config_barcode.tsv
>...
>```
> sample_sheet: ${RIBOVIZ_ORGANISM_FILES}/annotation/barcodes/Gupta_2018_tRNA_Modification_Carbon_Nitrogen_Metabolism_RPF_9-samples_CDS_w_250utrs_config_barcode.tsv 
>```

`sample_sheet` wouldn't be usable with the environment variables since the Nextflow workflow, in common with the Python one, assumes that `sample_sheet` is relative to `dir_in`.

**Comment by mikej888** [comment]
Up to commit ecbcac4 on [yaml-environment-vars-274](https://github.com/riboviz/riboviz/tree/yaml-environment-vars-274)

Updated `riboviz/utils.py`:

* Added `replace_tokens(string, tokens={})`

Updated `riboviz/test/test_utils.py`

* Added `test_replace_tokens`.

Added `riboviz/environment.py` with:

* `get_environment_vars()`
* `apply_env_to_config(config)`

Updated `riboviz/params.py`,

* Added:

```
ENV_RIBOVIZ_SAMPLES
ENV_RIBOVIZ_ORGANISMS
ENV_RIBOVIZ_DATA
```

* Added `ENV_PARAMS` defining list of all riboviz.params constants that can incl
ude environment variables.

Updated `riboviz/test/nextflow/__init__.py`

* `run_nextflow(config_file)` changed to `run_nextflow(config_file, envs={}, val
idate_only=False)`
 
Updated `riboviz/test/nextflow/test_nextflow_errors.py`:

* Renamed to `riboviz/test/nextflow/test_nextflow_validation.py`.
* Added `test_environment_vars(tmpdir)`
* Added `test_environment_var_not_found(tmpdir, env)`

Updated `riboviz/test/regression/conftest.py`:

* Updated `pytest_generate_tests(metafunc)` to replace environment variable toke
ns with environment variables in configuration parameter values that support env
ironment variables, via use of `environment.apply_env_to_config(config)`.

Updated `riboviz/test/regression/test_regression.py`:

* Updated `prep_riboviz_fixture(...)` to use `riboviz.test.nextflow.run_nextflow`.

**Comment by mikej888** [comment]
Rather than traverse and replace tokens in all parameters, for now `prep_riboviz.nf` handles each file-specific parameter only:
```
asite_disp_length_file
codon_positions_file
dir_in
dir_index
dir_out
dir_tmp
features_file
orf_fasta_file
orf_gff_file
rrna_fasta_file
t_rna_file
```
The following parameters refer to files relative to the above so don't need any token replacement to be done:
```
fq_files
multiplex_fq_files
orf_index_prefix
sample_sheet
rrna_index_prefix
```
There was no need to edit the section that creates a new YAML file for use by the `countReads` process for passing onto `count_reads.py`. This YAML only includes `fq_files` or `multiplex_fq_files` which are relative to `dir_in`, `dir_out` and `dir_tmp`, the latter of which which are provided as arguments to `count_reads.py`.

**Comment by mikej888** [comment]
Up to commit 92cc4da on branch [yaml-environment-vars-274](https://github.com/riboviz/riboviz/tree/yaml-environment-vars-274) has documentation updates:

[prep-riboviz-config.md](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/user/prep-riboviz-config.md)

* [Configuring file paths and directories (Nextflow workflow only)](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/user/prep-riboviz-config.md#configuring-file-paths-and-directories-nextflow-workflow-only): new section.
* [Environment variables and configuration tokens](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/user/prep-riboviz-config.md#environment-variables-and-configuration-tokens): new subsection.
* [Symbolic links and creating expected directory structures](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/user/prep-riboviz-config.md#symbolic-links-and-creating-expected-directory-structures). Subsection relocated from `prep-riboviz-run-nextflow.md`.

[prep-riboviz-run-nextflow.md](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/user/prep-riboviz-run-nextflow.md)

* [Run the Nextflow workflow](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/user/prep-riboviz-run-nextflow.md#run-the-nextflow-workflow). Section updated with information on how to provide environment variables on the command-line.

[testing.md](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/developer/testing.md)

* [Run vignette regression test suite](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/developer/testing.md#run-vignette-regression-test-suite). Section updated with information on how to provide environment variables on the command-line.
* Other sections rewritten for clarity.

**Comment by mikej888** [comment]
Note about what needs to happen when this branch is eventually merged...

Once merged into `develop` all others working on `prep_riboviz.nf` in their branches will need to replace occurrences of `params.<PARAM>` with `<PARAM>` for where `<PARAM>` is one of 
```
asite_disp_length_file
codon_positions_file
dir_in
dir_index
dir_out
dir_tmp
features_file
orf_fasta_file
orf_gff_file
rrna_fasta_file
t_rna_file
```
This replacement needs to be done in any processes or Groovy code that has been added in these branches. For example `params.dir_in` needs replaced by `dir_in`. This is because during intitialisation and parameter validation, these `params` values are extracted and have their environment variable tokens, if present, replaced. Unfortunately `params` cannot be updated in-place, hence the introduction of new variables.

Once this branch is merged, @FlicAnderson, for Rewrite diagnostic plots in generate_stats_figs.R as .Rmd (R markdown file) #193, should update code to explicitly create a new YAML file for `AnalysisOutputs.Rmd`, rather than dumping `params`, so that the parameters processed above can be inserted into the YAML file (also this saves `AnalysisOutputs.Rmd` having to worry about replacing environment variables in its configuration) Given that some of the parameters are optional I think something like the following will suffice:

```
Map analysis_outputs_inputs = [:]
if (is_asite_disp_length_file) {
    analysis_outputs_inputs.asite_disp_length_file
}
if (is_codon_positions_file) {
    analysis_outputs_inputs.codon_positions_file
}
if (is_features_file) {
    analysis_outputs_inputs.features_file
}
if (is_t_rna_file) {
    analysis_outputs_inputs.t_rna_file
}
analysis_outputs_inputs_yaml = new Yaml().dump(analysis_outputs_inputs_yaml)
```
(assuming that's the only parameters "AnalysisOutputs.Rmd` uses (if it needs more then these can be added to `analysis_outputs_inputs` also).

**Comment by mikej888** [comment]
Create configuration files with environment variables:
```console
$ cp vignette/vignette_config.yaml vignette/vignette_config_env.yaml
$ cp vignette/simdata_umi_config.yaml vignette/simdata_umi_config_env.yaml 
$ cp vignette/simdata_multiplex_config.yaml vignette/simdata_multiplex_config_env.yaml 
```
Update `vignette/vignette_config_env.yaml`:
```
asite_disp_length_file: ${RIBOVIZ_DATA}/yeast_standard_asite_disp_length.txt # Table of fixed A-site positions by read length
codon_positions_file: ${RIBOVIZ_DATA}/yeast_codon_pos_i200.RData # Codon positions in each gene
dir_index: ${RIBOVIZ_SAMPLES}/index # Built indices directory
dir_in: ${RIBOVIZ_SAMPLES}/input # Input directory
dir_logs: ${RIBOVIZ_SAMPLES}/logs # Log files directory
dir_out: ${RIBOVIZ_SAMPLES}/output # Output directory
dir_tmp: ${RIBOVIZ_SAMPLES}/tmp # Intermediate files directory
features_file: ${RIBOVIZ_DATA}/yeast_features.tsv # Features to correlate with ORFs
orf_fasta_file: ${RIBOVIZ_ORGANISMS}/yeast_YAL_CDS_w_250utrs.fa # ORF file to align to
orf_gff_file: ${RIBOVIZ_ORGANISMS}/yeast_YAL_CDS_w_250utrs.gff3 # GFF2/GFF3 file for ORFs
rrna_fasta_file: ${RIBOVIZ_ORGANISMS}/yeast_rRNA_R64-1-1.fa # rRNA file to avoid aligning to
t_rna_file: ${RIBOVIZ_DATA}/yeast_tRNAs.tsv # tRNA estimates
```
Update `vignette/simdata_umi_config_env.yaml`:
```
asite_disp_length_file: ${RIBOVIZ_DATA}/yeast_standard_asite_disp_length.txt # Table of fixed A-site positions by read length
dir_in: ${RIBOVIZ_SAMPLES}/simdata/ # Input directory
dir_index: ${RIBOVIZ_SAMPLES}/index # Built indices directory
dir_logs: ${RIBOVIZ_SAMPLES}/simdata_umi_logs # Log files directory
dir_out: ${RIBOVIZ_SAMPLES}/simdata_umi_output # Output directory
dir_tmp: ${RIBOVIZ_SAMPLES}/simdata_umi_tmp # Intermediate files directory
features_file: ${RIBOVIZ_DATA}/yeast_features.tsv # Features to correlate with ORFs
orf_fasta_file: ${RIBOVIZ_ORGANISMS}/yeast_YAL_CDS_w_250utrs.fa # ORF file to align to
orf_gff_file: ${RIBOVIZ_ORGANISMS}/yeast_YAL_CDS_w_250utrs.gff3 # GFF2/GFF3 file for ORFs
rrna_fasta_file: ${RIBOVIZ_ORGANISMS}/yeast_rRNA_R64-1-1.fa # rRNA file to avoid aligning to
```
Note: `dir_index`, `dir_tmp`, `dir_out` will be written to `${RIBOVIZ_SAMPLES}` i.e. `data/`.

Update `vignette/simdata_multiplex_config_env.yaml`:
```
asite_disp_length_file: ${RIBOVIZ_DATA}/yeast_standard_asite_disp_length.txt # Table of fixed A-site positions by read length
dir_in: ${RIBOVIZ_SAMPLES}/simdata/ # Input directory
dir_index: ${RIBOVIZ_SAMPLES}/index # directory to put built indices in
dir_logs: ${RIBOVIZ_SAMPLES}/simdata_multiplex_logs # Log files directory
dir_out: ${RIBOVIZ_SAMPLES}/simdata_multiplex_output # Output directory
dir_tmp: ${RIBOVIZ_SAMPLES}/simdata_multiplex_tmp # Intermediate files directory
features_file: ${RIBOVIZ_DATA}/yeast_features.tsv # Features to correlate with ORFs
orf_fasta_file: ${RIBOVIZ_ORGANISMS}/yeast_YAL_CDS_w_250utrs.fa # ORF file to align to
orf_gff_file: ${RIBOVIZ_ORGANISMS}/yeast_YAL_CDS_w_250utrs.gff3 # GFF2/GFF3 file for ORFs
rrna_fasta_file: ${RIBOVIZ_ORGANISMS}/yeast_rRNA_R64-1-1.fa # rRNA file to avoid aligning to
```
Note: `dir_index`, `dir_tmp`, `dir_out` will be written to `${RIBOVIZ_SAMPLES}` i.e. `data/`.

Run within `riboviz` directory:
```console
$ RIBOVIZ_SAMPLES=vignette/ \
RIBOVIZ_ORGANISMS=vignette/input/ \
RIBOVIZ_DATA=data \
nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config_env.yaml

$ RIBOVIZ_SAMPLES=data \
RIBOVIZ_ORGANISMS=vignette/input \
RIBOVIZ_DATA=data \
nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/simdata_umi_config_env.yaml

$ RIBOVIZ_SAMPLES=data \
RIBOVIZ_ORGANISMS=vignette/input \
RIBOVIZ_DATA=data \
nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/simdata_multiplex_config_env.yaml
```
```console
$ ls -1F data vignette
data:
...
index/
...
simdata_multiplex_output/
simdata_multiplex_tmp/
simdata_umi_output/
simdata_umi_tmp/
...
vignette:
...
index/
output/
tmp/
...
```
```console
$ RIBOVIZ_SAMPLES=vignette \
RIBOVIZ_ORGANISMS=vignette/input \
RIBOVIZ_DATA=data \
pytest riboviz/test/regression/test_regression.py --expected=vignette-orig-nf \
--config-file=vignette/vignette_config_env.yaml --skip-workflow --check-index-tmp --nextflow

$ RIBOVIZ_SAMPLES=data \
RIBOVIZ_ORGANISMS=vignette/input \
RIBOVIZ_DATA=data \
pytest riboviz/test/regression/test_regression.py --expected=vignette-orig-nf \
--config-file=vignette/simdata_umi_config_env.yaml --skip-workflow --check-index-tmp --nextflow

$ RIBOVIZ_SAMPLES=data \
RIBOVIZ_ORGANISMS=vignette/input \
RIBOVIZ_DATA=data \
pytest riboviz/test/regression/test_regression.py --expected=vignette-orig-nf \
--config-file=vignette/simdata_multiplex_config_env.yaml --skip-workflow --nextflow
```

**Comment by mikej888** [comment]
Example using custom directory structure but same YAML configuration files as above:

```console
$ cd
$ mkdir -p env-example-dataset/data
$ cp riboviz/data/yeast_codon_pos_i200.RData env-example-dataset/data/
$ cp riboviz/data/yeast_features.tsv env-example-dataset/data/
$ cp riboviz/data/yeast_standard_asite_disp_length.txt env-example-dataset/data/
$ cp riboviz/data/yeast_tRNAs.tsv env-example-dataset/data/

$ mkdir -p env-example-dataset/organisms/
$ cp riboviz/vignette/input/yeast_rRNA_R64-1-1.fa env-example-dataset/organisms/
$ cp riboviz/vignette/input/yeast_YAL_CDS_w_250utrs.fa env-example-dataset/organisms/
$ cp riboviz/vignette/input/yeast_YAL_CDS_w_250utrs.gff3 env-example-dataset/organisms/

$ mkdir -p env-example-dataset/samples
$ mkdir -p env-example-dataset/samples/input
$ cp riboviz/vignette/input/SRR1042855_s1mi.fastq.gz env-example-dataset/samples/input/
$ cp riboviz/vignette/input/SRR1042864_s1mi.fastq.gz env-example-dataset/samples/input/
$ mkdir -p env-example-dataset/samples/simdata
$ cp riboviz/data/simdata/umi5_umi3_umi_adaptor.fastq env-example-dataset/samples/simdata/
$ cp riboviz/data/simdata/multiplex_umi_barcode_adaptor.fastq env-example-dataset/samples/simdata/
$ cp riboviz/data/simdata/multiplex_barcodes.tsv env-example-dataset/samples/simdata/
```
```console
$ ls -1R env-example-dataset/
env-example-dataset/:
data
organisms
samples

env-example-dataset/data:
yeast_codon_pos_i200.RData
yeast_features.tsv
yeast_standard_asite_disp_length.txt
yeast_tRNAs.tsv

env-example-dataset/organisms:
yeast_rRNA_R64-1-1.fa
yeast_YAL_CDS_w_250utrs.fa
yeast_YAL_CDS_w_250utrs.gff3

env-example-dataset/samples:
input
simdata

env-example-dataset/samples/input:
SRR1042855_s1mi.fastq.gz
SRR1042864_s1mi.fastq.gz

env-example-dataset/samples/simdata:
multiplex_barcodes.tsv
multiplex_umi_barcode_adaptor.fastq
umi5_umi3_umi_adaptor.fastq
```

Run within `riboviz` using template configuration files and environment variables:

```console
$ RIBOVIZ_SAMPLES=~/env-example-dataset/samples \
RIBOVIZ_ORGANISMS=~/env-example-dataset/organisms \
RIBOVIZ_DATA=~/env-example-dataset/data \
nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config_env.yaml

$ RIBOVIZ_SAMPLES=~/env-example-dataset/samples \
RIBOVIZ_ORGANISMS=~/env-example-dataset/organisms \
RIBOVIZ_DATA=~/env-example-dataset/data \
nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/simdata_umi_config_env.yaml

$ RIBOVIZ_SAMPLES=~/env-example-dataset/samples \
RIBOVIZ_ORGANISMS=~/env-example-dataset/organisms \
RIBOVIZ_DATA=~/env-example-dataset/data \
nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/simdata_multiplex_config_env.yaml
```
```console
$ ls -1F vignette data/
```
No index, temporary or output files, as expected.
```console
$ ls -1F ~/env-example-dataset/samples/
index/
input/
output/
simdata/
simdata_multiplex_output/
simdata_multiplex_tmp/
simdata_umi_output/
simdata_umi_tmp/
tmp/
```
Index, temporary and output files, as expected.
```console
$ RIBOVIZ_SAMPLES=~/env-example-dataset/samples/ \
RIBOVIZ_ORGANISMS=~/env-example-dataset/organisms/ \
RIBOVIZ_DATA=~/env-example-dataset/data/ \
pytest riboviz/test/regression/test_regression.py --expected=vignette-orig-nf \
--config-file=vignette/vignette_config_env.yaml --skip-workflow --check-index-tmp --nextflow

$ RIBOVIZ_SAMPLES=~/env-example-dataset/samples/ \
RIBOVIZ_ORGANISMS=~/env-example-dataset/organisms/ \
RIBOVIZ_DATA=~/env-example-dataset/data/ \
pytest riboviz/test/regression/test_regression.py --expected=../riboviz/vignette-orig-nf \
--config-file=vignette/simdata_umi_config_env.yaml --skip-workflow --check-index-tmp --nextflow

$ RIBOVIZ_SAMPLES=~/env-example-dataset/samples/ \
RIBOVIZ_ORGANISMS=~/env-example-dataset/organisms/ \
RIBOVIZ_DATA=~/env-example-dataset/data/ \
pytest riboviz/test/regression/test_regression.py --expected=../riboviz/vignette-orig-nf \
--config-file=vignette/simdata_multiplex_config_env.yaml --skip-workflow --nextflow
```

**Comment by mikej888** [comment]
TODO:

* Write tests for `riboviz.environment`
* Write tests in `riboviz.tests.nextflow` for workflow configurations that use environment variables.

**Comment by mikej888** [comment]
@ewallace @FlicAnderson 
While developing the tests I've found an issue. I cannot come up with a configuration has environment variable tokens such that it works for the same directory structure both in the presence of environment variables and in their absence (where the default of `.` is used).
For example, given a directory structure:
```
example/
  samples/
   input/
  organisms/
  data/
```
and parameters:
```
dir_in: ${RIBOVIZ_SAMPLES}/input
asite_disp_length_file: ${RIBOVIZ_DATA}/yeast_standard_asite_disp_length.txt
```
Running the following command anywhere:
```console
$ RIBOVIZ_SAMPLES=<path>/example/samples nextflow run ...
```
succeeds. `dir_in` becomes `<path>/example/samples/input` and `asite_disp_length_file` becomes `<path>/example/data/yeast_standard_asite_disp_length.txt`.

But, running the following command within `example`:
```console
$ nextflow run ...
```
fails. `dir_in` becomes `./input` and `asite_disp_length_file` becomes `./yeast_standard_asite_disp_length.txt`.
So the cost may be - if you want to use a configuration that has environment variable tokens you _must_ provide the environment variable values, and not rely on relative defaults. To enforce that, the notion of defaults could be removed. Instead the workflow could raise an error that tokens with environment variables have been found for which no environment variables have been declared.

**Comment by mikej888** [comment]
Commit 9c60525 of [yaml-environment-vars-274](https://github.com/riboviz/riboviz/tree/yaml-environment-vars-274) has `riboviz.test.nextflow.test_configuration`. This has tests for configurations with environment variable tokens (both in the presence and absence of the associated environment variables), and configurations with relative paths and absolute paths. For each test configurations based on all of `vignette_config.yaml`, `vignette_simdata_umi_config.yaml` and `vignette_simdata_multiplex_config.yaml` are run. Tests currently run `nextflow run ... --validate_only` which is fine for checking that input files are handled OK.

Tests are needed to run the workflow to ensure that index, temporary and output file paths, which can also have environment variable tokens, are created OK. Depending on how rigourous the test coverage is to be this will incur running `nextflow` on both each of `vignette_simdata_umi_config.yaml` and `vignette_simdata_multiplex_config.yaml` for each of the following combinations:

* Config with environment variable tokens in the presence of the associated environment variables. 
* Config with environment variable tokens in the absence of the associated environment variables.
* Config with relative paths.
* Config with absolute path.

i.e. 8 runs of the simulated data workflows in total.

**Comment by mikej888** [comment]
Commit 4db7b48 of [yaml-environment-vars-274](https://github.com/riboviz/riboviz/tree/yaml-environment-vars-274):

Updates to `riboviz.test.nextflow.test_configuration`:

* Test functions replaced by a single test function which is parameterized with test cases.
* Each test case includes references to callback functions to create test directory, get configuration, get environment variables, get directory in which Nextflow is to be run.
* The test function, `test_run`, validates the workflow, and, optionally, depending upon the test configuration, runs the workflow and checks that the temporary and output directories have been created and `TPMS_collated.tsv` exists (the contents are not validated as these tests are designed to test the configuration paths only).

The test function is parameterised by both test cases and configurations:

| Configuration paths | Environment variable tokens | Environment variables defined? | Input files location | Workflow run location |
| - | - | - | - | - |
| Default | No | N/A | `riboviz` | `riboviz` |
| Default | No | N/A | Test directory | Test directory |
| Use environment variable tokens | Yes | Yes | Test directory | `riboviz` |
| Use environment variable tokens | Yes | No (defaukt, `.`, used) | Test directory | Test directory |
| Relative | No | N/A | Test directory | Test directory |
| Absolute | No | N/A | Test directory | `riboviz` |

| Configuration | Validation only? |
| - | - |
| `vignette/vignette_config.yaml` | True |
| `vignette/simdata_umi_config.yaml` | False |
| `vignette/simdata_multiplex_config.yaml`  | True |

In total, 18 tests are run, 12 that do validation only and 6 that run the workflow

**Comment by mikej888** [comment]
To be discussed:

* Sample sheet continues to be relative to `dir_in`. See https://github.com/riboviz/riboviz/issues/274#issuecomment-796708633
* Requirements for those working on other branches when [yaml-environment-vars-274](https://github.com/riboviz/riboviz/tree/yaml-environment-vars-274) is merged. See https://github.com/riboviz/riboviz/issues/274#issuecomment-802054848
* Recommendation that, if the configuration file contains environment variables, then the user must provide values for these. See https://github.com/riboviz/riboviz/issues/274#issuecomment-805085412.

**Comment by ewallace** [comment]
How to test:
1. get data and files and put them in 2 places
2. check that riboviz runs
3. file any bug reports or fixes, or edits to documentation in [prep-riboviz-config.md](https://github.com/riboviz/riboviz/blob/yaml-environment-vars-274/docs/user/prep-riboviz-config.md#configuring-file-paths-and-directories-nextflow-workflow-only)

**Comment by ewallace** [comment]
## Testing on tinysim

tl;dr everything worked eventually, but I had some wrong turns through misunderstanding environment variables, I think can be addressed in documentation.

### Configuration file

I modified the configuration file [Mok-tinysim_config.yaml](https://github.com/riboviz/example-datasets/blob/master/simulated/mok/Mok-tinysim_config.yaml) changing the lines below (in branch `test-envt-274`)

```
asite_disp_length_file: ${RIBOVIZ_DATA}/yeast_standard_asite_disp_length.txt # Table of fixed A-site positions by read length
...
dir_in: ${RIBOVIZ_SAMPLES}/input # Input directory
dir_index: ${RIBOVIZ_SAMPLES}/index # Built indices directory
dir_logs: ${RIBOVIZ_SAMPLES}/logs # Log files directory
dir_out: ${RIBOVIZ_SAMPLES}/output # Output directory
dir_tmp: ${RIBOVIZ_SAMPLES}/tmp # Intermediate files directory
...
orf_fasta_file: ${RIBOVIZ_ORGANISMS}/annotation/tiny_2genes_20utrs.fa # ORF file to align to
orf_gff_file: ${RIBOVIZ_ORGANISMS}/annotation/tiny_2genes_20utrs.gff3 # GFF2/GFF3 file for ORFs
...
rrna_fasta_file: ${RIBOVIZ_ORGANISMS}/contaminants/Sc_rRNA_example.fa # rRNA file to avoid aligning to
...
```

#### Set up data

I have a `/home/$USER/riboviz/` directory, inside which the `riboviz` and `example-datasets` repos live.

First created directories on my Eddie scratch space (while logged on to Eddie):
```
mkdir /exports/eddie/scratch/$USER/tinysim
mkdir /exports/eddie/scratch/$USER/tinysim/input
```

Then uploaded the data from [simRiboSeq/simulation_runs/riboviz](https://github.com/amandamok/simRiboSeq/tree/master/simulation_runs/riboviz) from my local machine:
```
UUN=ewallac2
scp tiny_2genes.fq $UUN@eddie.ecdf.ed.ac.uk:/exports/eddie/scratch/$UUN/tinysim/input
```

#### Try 1: wrong environment in call

Then I tried to follow the documentation, but naively with environment variables specified on previous lines:
```
RIBOVIZ_SAMPLES=/exports/eddie/scratch/$USER/tinysim
RIBOVIZ_DATA=/home/$USER/riboviz/riboviz/data
RIBOVIZ_ORGANISMS=/home/$USER/riboviz/example-datasets/simulated/mok
nextflow prep_riboviz.nf -params-file ${RIBOVIZ_ORGANISMS}/Mok-tinysim_config.yaml
```

This gave the error:
```
N E X T F L O W  ~  version 20.04.1
Launching `prep_riboviz.nf` [friendly_golick] - revision: 1cbd4fe407
samples_dir: .
organisms_dir: .
data_dir: .
No such directory (dir_in): ./input
```

#### Try 2: wrong environment in `params-file`

Following the documentation more precisely, in a new shell session I tried specifying the environment variables in the same line (with `\` line continuation):
```
RIBOVIZ_SAMPLES=/exports/eddie/scratch/$USER/tinysim \
RIBOVIZ_DATA=/home/$USER/riboviz/riboviz/data \
RIBOVIZ_ORGANISMS=/home/$USER/riboviz/example-datasets/simulated/mok \
nextflow prep_riboviz.nf -params-file ${RIBOVIZ_ORGANISMS}/Mok-tinysim_config.yaml
```

This failed because `${RIBOVIZ_ORGANISMS}` was not in the right environment for the nextflow call (only, I think, inside the nextflow call):
```
N E X T F L O W  ~  version 20.04.1
Launching `prep_riboviz.nf` [fabulous_volhard] - revision: 1cbd4fe407
Specified params file does not exists: /Mok-tinysim_config.yaml
```

#### Try 3: success

Next, I tried:
```
RIBOVIZ_SAMPLES=/exports/eddie/scratch/$USER/tinysim \
RIBOVIZ_DATA=/home/$USER/riboviz/riboviz/data \
RIBOVIZ_ORGANISMS=/home/$USER/riboviz/example-datasets/simulated/mok \
nextflow prep_riboviz.nf -params-file \
/home/$USER/riboviz/example-datasets/simulated/mok/Mok-tinysim_config.yaml
```

This ran perfectly:
```
N E X T F L O W  ~  version 20.04.1
Launching `prep_riboviz.nf` [distraught_visvesvaraya] - revision: 1cbd4fe407
samples_dir: /exports/eddie/scratch/ewallac2/tinysim
organisms_dir: /home/ewallac2/riboviz/example-datasets/simulated/mok
data_dir: /home/ewallac2/riboviz/riboviz/data
executor >  local (14)
executor >  local (14)
[ab/daa63a] process > buildIndicesrRNA (yeast_rRNA)        [100%] 1 of 1 ✔
[cd/b76451] process > buildIndicesORF (tiny_2genes_20utrs) [100%] 1 of 1 ✔
[09/8f5e0c] process > cutAdapters (A)                      [100%] 1 of 1 ✔
[-        ] process > extractUmis                          -
[-        ] process > cutAdaptersMultiplex                 -
[-        ] process > extractUmisMultiplex                 -
[-        ] process > demultiplex                          -
[59/779ca8] process > hisat2rRNA (A)                       [100%] 1 of 1 ✔
[ca/b4615a] process > hisat2ORF (A)                        [100%] 1 of 1 ✔
[da/520b58] process > trim5pMismatches (A)                 [100%] 1 of 1 ✔
[aa/ed8446] process > samViewSort (A)                      [100%] 1 of 1 ✔
[-        ] process > groupUmisPreDedup                    -
[-        ] process > dedupUmis                            -
[-        ] process > groupUmisPostDedup                   -
[20/bba1f3] process > outputBams (A)                       [100%] 1 of 1 ✔
[f2/60dcdf] process > makeBedgraphs (A)                    [100%] 1 of 1 ✔
[2f/9832c2] process > bamToH5 (A)                          [100%] 1 of 1 ✔
[ea/d161af] process > generateStatsFigs (A)                [100%] 1 of 1 ✔
[f7/3c38f1] process > renameTpms (A)                       [100%] 1 of 1 ✔
[92/773ccb] process > collateTpms (A)                      [100%] 1 of 1 ✔
[c1/429e2b] process > countReads                           [100%] 1 of 1 ✔
Finished processing sample: A
```

I next checked that `output/A/plus.bedgraph` was as expected by comparing with `data/Mok-tinysim-gffsam/A.sam`. and so on.

## Conclusion.

I conclude that the environment variables definitely work. But we have to document the usage better for unix non-experts like myself. We should be explicit in the documentation about specifying the environment variables  *on the same line before the nextflow call*.

**Comment by ewallace** [comment]
#### T
First I used `fasterq-dump` and then `pigz` to place two sample files in `/exports/eddie/scratch/$USER/L-Sc2014`.

Edited the config file [Lareau_2014_Replicates_RPF_3-samples_CDS_w_250utrs_config.yaml](/riboviz/example-datasets/fungi/saccharomyces/Lareau_2014_Replicates_RPF_3-samples_CDS_w_250utrs_config.yaml) to change only:
```
dir_in: ${RIBOVIZ_SAMPLES}/input # Input directory
dir_index: ${RIBOVIZ_SAMPLES}/index # Built indices directory
dir_logs: ${RIBOVIZ_SAMPLES}/logs # Log files directory
dir_out: ${RIBOVIZ_SAMPLES}/output # Output directory
dir_tmp: ${RIBOVIZ_SAMPLES}/tmp # Intermediate files directory
...
orf_fasta_file: ${RIBOVIZ_ORGANISMS}/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.fa # ORF file to$
orf_gff_file: ${RIBOVIZ_ORGANISMS}/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.gff3 # GFF2/GFF3 f$
...
rrna_fasta_file: ${RIBOVIZ_ORGANISMS}/contaminants/Saccharomyces_cerevisiae_yeast_rRNA_R64-1-1.fa # rRNA file$
...
```
I also commented out the 3rd sample, concerned about space.

Then I ran:
```
RIBOVIZ_ORGANISMS=$HOME/riboviz/example-datasets/fungi/saccharomyces \
RIBOVIZ_SAMPLES=/exports/eddie/scratch/$USER/L-Sc2014 \
nextflow prep_riboviz.nf -params-file \
$HOME/riboviz/example-datasets/fungi/saccharomyces/Lareau_2014_Replicates_RPF_3-samples_CDS_w_250utrs_config.yaml \
-with-report L-Sc2014_test-envt-274_2021-05-19.html
```

This launched off correctly, indicating that riboviz was able to find the annotation files and data correctly as specified in the environment variables:
```
N E X T F L O W  ~  version 20.04.1
Launching `prep_riboviz.nf` [big_galileo] - revision: 1cbd4fe407
samples_dir: /exports/eddie/scratch/ewallac2/L-Sc2014
organisms_dir: /home/ewallac2/riboviz/example-datasets/fungi/saccharomyces
data_dir: .
executor >  local (4)
[39/6bf6bb] process > buildIndicesrRNA (yeast_rRNA)     [100%] 1 of 1 ✔
[6e/f11cd6] process > buildIndicesORF (yeast_CDS_w_250) [100%] 1 of 1 ✔
[b3/2089aa] process > cutAdapters (Replicate-1)         [  0%] 0 of 2
...
```

I may update this if it finishes. It looks great so far.

**Comment by ewallace** [comment]
I think we can come up with better names for these tokens. So far we have:
- `RIBOVIZ_SAMPLES`
- `RIBOVIZ_ORGANISMS`
- `RIBOVIZ_DATA`

The problem is that it's all data. In `example-datasets` we have the organism-specific data, which is currently broken up into `annotation` and `contaminant` folders. Really they are similar, it is organism-specific fasta and gff files. Then there is some data that we have been keeping in this repository `data`, but most of that is yeast-specific data that is there out of convenience and doesn't belong, I mean codon positions and tRNA concentrations and so on. It "belongs"  with the organism-specific data in `example-datasets`. That might change if we calculate some things like codon position matrix directly from the annotation.

Then the input fastq files (and sample sheet) are a different kind of thing, and different from the output. For the vignette, we have them in `vignette/input` and then have `vignette/output`, but I don't think they are necessarily linked.

So I am wondering about variables more like
- `RIBOVIZ_SAMPLEINPUT` - for fastq files and sample sheet, a bit like `dir_in`
- `RIBOVIZ_ORGANISMS` - for all organism-specific data that we want to organize like this
- `RIBOVIZ_OTHERDATA` - maybe, for other kinds of input data like the `asite_disp_length_file`
- `RIBOVIZ_OUTPUT` - simplify path output directory which is likely to be distinct from the sample input.

I may be making it too complicated. It would be good to collect ideas from a few other users.

**Comment by FlicAnderson** [comment]
@ewallace I think `RIBOVIZ_SAMPLEINPUT` shouldn't cover fastq files and the sample sheet? Or if it does, we need to think about re-jigging example-datasets' structure?  

Thinking about a user trying out an example dataset from our repo: users will be downloading their fastq files to somewhere which won't be the same location as sample sheets (which would be in the genus-level folder, e.g. for the Gupta datasets, it's in saccharomyces). This means the user then has to move either the FASTQs they've downloaded into that folder, which gets messy, OR move the sample sheet out, which seems to undo some of the usefulness of a 'ready-to-use' example dataset? 

Does this make sense?

**Comment by FlicAnderson** [comment]
@ewallace I think shaking example-datasets up a bit to have 'organism-specific' and perhaps 'dataset-specific' folders could be an interesting idea - for example annotations/contaminants .fa, .gff, provenances in organism-specific folder as you mentioned; sample-sheets, ?config files? potentially in a dataset-specific folder?  Perhaps that's too much?  Definitely not in the scope of this ticket though! :)

I am slightly resistant to `_OTHERDATA` because it seems less helpful than `_DATA`? In order to know what goes in OTHERDATA, you need to know all the other variable names and have a good idea what goes into those? DATA seems also a bit vague, but I would need to try this way of working out and test it before I can give you any better alternatives, which isn't likely to happen too soon I'm afraid! :disappointed:

**Comment by mikej888** [comment]
How about we proceed by merging the work to date then create a new issue with your suggestions above that we can review after others have used this feature also? The Nextflow code that handles the replacements is located within a dozen lines within the workflow so is straightforward to rename variables and/or add support for more at a later date (similarly for the documentation and tests).

**Comment by FlicAnderson** [comment]
@mikej888 I agree. Testing and merging, then tweaking on subsequent tickets sounds like a good plan.

**Comment by mikej888** [comment]
@ewallace, for your issues with setting variables. For:
```console
$ RIBOVIZ_SAMPLES=/exports/eddie/scratch/$USER/tinysim
$ RIBOVIZ_DATA=/home/$USER/riboviz/riboviz/data
$ RIBOVIZ_ORGANISMS=/home/$USER/riboviz/example-datasets/simulated/mok
$ nextflow prep_riboviz.nf -params-file ${RIBOVIZ_ORGANISMS}/Mok-tinysim_config.yaml
```
I replicated an analogous failure. It fails as the environment variable values are not available within Nextflow. I suspect that Nextflow runs, at least partly, within a subshell, and the above defines variables in the current shell only. 
For:
```console
$ RIBOVIZ_SAMPLES=/exports/eddie/scratch/$USER/tinysim \
  RIBOVIZ_DATA=/home/$USER/riboviz/riboviz/data \
  RIBOVIZ_ORGANISMS=/home/$USER/riboviz/example-datasets/simulated/mok \
  nextflow prep_riboviz.nf -params-file ${RIBOVIZ_ORGANISMS}/Mok-tinysim_config.yaml
```
This fails as `${RIBOVIZ_ORGANISMS}` in `${RIBOVIZ_ORGANISMS}/Mok-tinysim_config.yaml` is expanded in the current shell before the command is run (and before the variables are defined). So the command actually run is:
```
RIBOVIZ_SAMPLES=/exports/eddie/scratch/$USER/tinysim \
RIBOVIZ_DATA=/home/$USER/riboviz/riboviz/data \
RIBOVIZ_ORGANISMS=/home/$USER/riboviz/example-datasets/simulated/mok \
nextflow prep_riboviz.nf -params-file /Mok-tinysim_config.yaml
```
This can be seen with this example, where `TEST` is undefined:
```console
$ echo $TEST
$ TEST=1 echo ${TEST}
```
Nothing is printed as the command run is `echo`. Compare this with:
```console
$ TEST=1 env | grep TEST
TEST=1
```
where `env` prints the environment within which the command is running, which shows that which `TEST` is available with the expected value.
A solution to both of these issues is to define the environment variables using `export` so they are available in sub-shells. For example:
```
$ export RIBOVIZ_SAMPLES=riboviz/vignette/input
$ export RIBOVIZ_ORGANISMS=riboviz/vignette
$ export RIBOVIZ_DATA=riboviz/data
$ nextflow run riboviz/prep_riboviz.nf -ansi-log false -params-file ${RIBOVIZ_ORGANISMS}/vignette_config.yaml --validate_only
N E X T F L O W  ~  version 20.04.1
Launching `riboviz/prep_riboviz.nf` [desperate_wing] - revision: 1cbd4fe407
Validating configuration only
samples_dir: riboviz/vignette/input
organisms_dir: riboviz/vignette
data_dir: riboviz/data
...
```
So for users options are:

* Define variables on the same line as the Nextflow invocation, but do not use any of those variables to specify the path to the configuration file.
* Define variables via `export`.
* Combinations of these.

I'll tweak the docs.

**Comment by mikej888** [comment]
@ewallace updates to docs

* b2694ef, updates prep_riboviz.nf comment, doc/developer/testing.md, docs/user/prep-riboviz-config.md, docs/user/prep-riboviz-run-nextflow.md, with clearer information about how to define environment variables.
* af9d349, updates docs/user/prep-riboviz-run-nextflow.md with the gotcha about using environment variable in path to `<CONFIG_FILE>`  and how user should either specify actual path or define environment variable using `export`.

**Comment by mikej888** [comment]
Closed as PR #298 was merged 2 weeks ago.
